<a href="https://colab.research.google.com/github/rishabhm4/DeepLearning-task/blob/master/Task2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

###############################################################################################################################
##ASSIGNMENT_02 _IDL


###############################################################################################################################

### task_01

In [0]:
import tensorflow as tf
import datetime
from datasets import MNISTDataset
from time import time


# get the data
mnist = tf.keras.datasets.mnist
(train_imgs, train_lbls), (test_imgs, test_lbls) = tf.keras.datasets.mnist.load_data()
mnist = MNISTDataset(train_imgs.reshape((-1, 784)), train_lbls,
                     test_imgs.reshape((-1, 784)), test_lbls,
                    batch_size=256, seed=int(time()))


# define the model first, from input to output

# this is a super deep model, cool!
n_units = 100
n_layers = 2
w_range = 0.4


# just set up a "chain" of hidden layers
layers = []
for layer in range(n_layers):
    layers.append(tf.keras.layers.Dense(
        n_units, activation=tf.nn.relu,
        kernel_initializer=tf.initializers.RandomUniform(minval=-w_range,
                                                         maxval=w_range),
        bias_initializer=tf.initializers.constant(0.001)))

# finally add the output layer
layers.append(tf.keras.layers.Dense(
    10, kernel_initializer=tf.initializers.RandomUniform(minval=-w_range,
                                                         maxval=w_range)))

lr = 0.1
for step in range(2000):
    img_batch, lbl_batch = mnist.next_batch()
    with tf.GradientTape() as tape:
        # here we just run all the layers in sequence via a for-loop
        out = img_batch
        for layer in layers:
            out = layer(out)
        xent = tf.reduce_mean(tf.nn.sparse_softmax_cross_entropy_with_logits(
            logits=out, labels=lbl_batch))

    weights = [var for l in layers for var in l.trainable_variables]
    grads = tape.gradient(xent, weights)
    for grad, var in zip(grads, weights):
        var.assign_sub(lr*grad)

    if not step % 100:
        preds = tf.argmax(out, axis=1, output_type=tf.int32)
        acc = tf.reduce_mean(tf.cast(tf.equal(preds, lbl_batch), tf.float32))
        print("Loss: {} Accuracy: {}".format(xent, acc))


out = mnist.test_data
for layer in layers:
    out = layer(out)
test_preds = tf.argmax(out, axis=1, output_type=tf.int32)
acc = tf.reduce_mean(tf.cast(tf.equal(test_preds, mnist.test_labels), tf.float32))
print("Final test accuracy: {}".format(acc))


# FAIL_01

##### Observations:
* The Loss function is leading to large value of "nan"
* Accuracy is lowest


#### Possible_fixes
* w_range =0.0001 gives minimum loss 
* with only "2" layers accuracy raises to 95-96 % and loss reaches



In [0]:
# then load/run tensorboard

%load_ext tensorboard

The tensorboard extension is already loaded. To reload it, use:
  %reload_ext tensorboard


In [0]:
%tensorboard --logdir logs

Reusing TensorBoard on port 6006 (pid 32648), started 1:37:06 ago. (Use '!kill 32648' to kill it.)






### task_02

In [0]:
import tensorflow as tf
import datetime
from datasets import MNISTDataset
from time import time


# get the data

(train_imgs, train_lbls), (test_imgs, test_lbls) = tf.keras.datasets.mnist.load_data()
mnist = MNISTDataset(train_imgs.reshape((-1, 784)), train_lbls,
                     test_imgs.reshape((-1, 784)), test_lbls,
                     batch_size=256, seed=int(time()))


# define the model first, from input to output

# this is a super deep model, cool!
n_units = 100
n_layers = 8
w_range = 0.1

# just set up a "chain" of hidden layers
layers = []
for layer in range(n_layers):
    layers.append(tf.keras.layers.Dense(
        n_units, activation=tf.nn.sigmoid,
        kernel_initializer=tf.initializers.RandomUniform(minval=-w_range,
                                                         maxval=w_range),
        bias_initializer=tf.initializers.constant(0.001)))

# finally add the output layer
layers.append(tf.keras.layers.Dense(
    10, kernel_initializer=tf.initializers.RandomUniform(minval=-w_range,
                                                         maxval=w_range)))

lr = 0.1
for step in range(2000):
    img_batch, lbl_batch = mnist.next_batch()
    with tf.GradientTape() as tape:
        # here we just run all the layers in sequence via a for-loop
        out = img_batch
        for layer in layers:
            out = layer(out)
        xent = tf.reduce_mean(tf.nn.sparse_softmax_cross_entropy_with_logits(
            logits=out, labels=lbl_batch))

    weights = [var for l in layers for var in l.trainable_variables]
    grads = tape.gradient(xent, weights)
    for grad, var in zip(grads, weights):
        var.assign_sub(lr*grad)

    if not step % 100:
        preds = tf.argmax(out, axis=1, output_type=tf.int32)
        acc = tf.reduce_mean(tf.cast(tf.equal(preds, lbl_batch), tf.float32))
        print("Loss: {} Accuracy: {}".format(xent, acc))


out = mnist.test_data
for layer in layers:
    out = layer(out)
test_preds = tf.argmax(out, axis=1, output_type=tf.int32)
acc = tf.reduce_mean(tf.cast(tf.equal(test_preds, mnist.test_labels), tf.float32))
print("Final test accuracy: {}".format(acc))


Loss: 2.302588701248169 Accuracy: 0.0546875
Loss: 2.3014566898345947 Accuracy: 0.0859375
Loss: 2.2999491691589355 Accuracy: 0.1328125
Starting new epoch...
Loss: 2.299797296524048 Accuracy: 0.125
Loss: 2.3034958839416504 Accuracy: 0.1015625
Starting new epoch...
Loss: 2.3005361557006836 Accuracy: 0.12109375
Loss: 2.296292304992676 Accuracy: 0.1328125
Loss: 2.2952463626861572 Accuracy: 0.12109375
Starting new epoch...
Loss: 2.2966763973236084 Accuracy: 0.1171875
Loss: 2.287877321243286 Accuracy: 0.13671875
Starting new epoch...
Loss: 2.2283103466033936 Accuracy: 0.19921875
Loss: 1.8669496774673462 Accuracy: 0.2734375
Starting new epoch...
Loss: 2.2528600692749023 Accuracy: 0.1796875
Loss: 1.3025857210159302 Accuracy: 0.45703125
Loss: 1.0507081747055054 Accuracy: 0.58984375
Starting new epoch...
Loss: 0.8838106393814087 Accuracy: 0.66796875
Loss: 0.8102004528045654 Accuracy: 0.7890625
Starting new epoch...
Loss: 0.4216696619987488 Accuracy: 0.88671875
Loss: 0.2864460349082947 Accuracy: 0

In [0]:
# then load/run tensorboard

%load_ext tensorboard
%tensorboard --logdir logs

The tensorboard extension is already loaded. To reload it, use:
  %reload_ext tensorboard


Reusing TensorBoard on port 6006 (pid 32648), started 14:27:50 ago. (Use '!kill 32648' to kill it.)

# FAIL_02


#### Observation
* loss function is shows very minimal change
* grads in hidden layer are convering to infinitesimally smaller values, vanishing
* activation function is sigmoidal



#### Possible_fixes
* activation function ReLu instead of sigmoidal, acc of 92% could be achieved









### task_03

In [0]:
import tensorflow as tf
import datetime 
from datasets import MNISTDataset
from time import time


# get the data
(train_imgs, train_lbls), (test_imgs, test_lbls) = tf.keras.datasets.mnist.load_data()
mnist = MNISTDataset(train_imgs.reshape((-1, 784)), train_lbls,
                     test_imgs.reshape((-1, 784)), test_lbls,
                     batch_size=256, seed=int(time()))


# define the model first, from input to output

# let's use fewer layers...
n_units = 100
n_layers = 2

# just set up a "chain" of hidden layers
layers = []
for layer in range(n_layers):
    layers.append(tf.keras.layers.Dense(
        n_units, activation=tf.nn.relu,
        kernel_initializer=tf.initializers.RandomUniform(minval=-0.01,
                                                         maxval=0.),
        bias_initializer=tf.initializers.constant(0.001)))

# finally add the output layer
layers.append(tf.keras.layers.Dense(
    10, kernel_initializer=tf.initializers.RandomUniform(minval=-0.01,
                                                         maxval=0.01)))

lr = 0.1
for step in range(2000):
    img_batch, lbl_batch = mnist.next_batch()
    with tf.GradientTape() as tape:
        # here we just run all the layers in sequence via a for-loop
        out = img_batch
        for layer in layers:
            out = layer(out)
        xent = tf.reduce_mean(tf.nn.sparse_softmax_cross_entropy_with_logits(
            logits=out, labels=lbl_batch))

    weights = [var for l in layers for var in l.trainable_variables]
    grads = tape.gradient(xent, weights)
    for grad, var in zip(grads, weights):
        var.assign_sub(lr*grad)

    if not step % 100:
        preds = tf.argmax(out, axis=1, output_type=tf.int32)
        acc = tf.reduce_mean(tf.cast(tf.equal(preds, lbl_batch), tf.float32))
        print("Loss: {} Accuracy: {}".format(xent, acc))


out = mnist.test_data
for layer in layers:
    out = layer(out)
test_preds = tf.argmax(out, axis=1, output_type=tf.int32)
acc = tf.reduce_mean(tf.cast(tf.equal(test_preds, mnist.test_labels), tf.float32))
print("Final test accuracy: {}".format(acc))


Loss: 2.302584409713745 Accuracy: 0.109375
Loss: 2.302786350250244 Accuracy: 0.08203125
Loss: 2.2919421195983887 Accuracy: 0.1875
Starting new epoch...
Loss: 2.3001952171325684 Accuracy: 0.0859375
Loss: 2.3054590225219727 Accuracy: 0.1015625
Starting new epoch...
Loss: 2.297451972961426 Accuracy: 0.16015625
Loss: 2.3041024208068848 Accuracy: 0.08984375
Loss: 2.3011035919189453 Accuracy: 0.109375
Starting new epoch...
Loss: 2.2965352535247803 Accuracy: 0.12890625
Loss: 2.3081459999084473 Accuracy: 0.07421875
Starting new epoch...
Loss: 2.2942192554473877 Accuracy: 0.13671875
Loss: 2.3003830909729004 Accuracy: 0.109375
Starting new epoch...
Loss: 2.3007216453552246 Accuracy: 0.0859375
Loss: 2.3001747131347656 Accuracy: 0.11328125
Loss: 2.2988152503967285 Accuracy: 0.1328125
Starting new epoch...
Loss: 2.3033738136291504 Accuracy: 0.078125
Loss: 2.305172920227051 Accuracy: 0.09765625
Starting new epoch...
Loss: 2.300079345703125 Accuracy: 0.13671875
Loss: 2.3026082515716553 Accuracy: 0.09

In [0]:
# then load/run tensorboard

%load_ext tensorboard
%tensorboard --logdir logs

The tensorboard extension is already loaded. To reload it, use:
  %reload_ext tensorboard


Reusing TensorBoard on port 6006 (pid 32648), started 14:32:04 ago. (Use '!kill 32648' to kill it.)

# FAIL_03

#### Observations
* weights are initialised in range of -0.1 to 0.
* we have used ReLu activation function, the hidden layer output becomes 0
* no weight updation 

#### Possible_fixes
* changing weights to range -0.1 to 0.1, ensures test acc of nearly 95%









### task_04

In [0]:
import tensorflow as tf
import datetime
from datasets import MNISTDataset
from time import time


# get the data
(train_imgs, train_lbls), (test_imgs, test_lbls) = tf.keras.datasets.mnist.load_data()
mnist = MNISTDataset(train_imgs.reshape((-1, 784)), train_lbls,
                     test_imgs.reshape((-1, 784)), test_lbls,
                     batch_size=256, seed=int(time()))


# define the model first, from input to output

# let's use fewer layers...
n_units = 100
n_layers = 2

# just set up a "chain" of hidden layers
layers = []
for layer in range(n_layers):
    layers.append(tf.keras.layers.Dense(
        n_units, activation=tf.nn.relu,
        kernel_initializer=tf.initializers.RandomUniform(minval=-0.01,
                                                         maxval=0.01),
        bias_initializer=tf.initializers.constant(0.001)))

# finally add the output layer
layers.append(tf.keras.layers.Dense(
    10, kernel_initializer=tf.initializers.RandomUniform(minval=-0.01,
                                                         maxval=0.01)))

lr = 0.1
for step in range(2000):
    img_batch, lbl_batch = mnist.next_batch()
    # I hear adding random noise to inputs helps with generalization!
    img_batch = img_batch + tf.random.normal(tf.shape(img_batch), stddev=1)
    with tf.GradientTape() as tape:
        # here we just run all the layers in sequence via a for-loop
        out = img_batch
        for layer in layers:
            out = layer(out)
        xent = tf.reduce_mean(tf.nn.sparse_softmax_cross_entropy_with_logits(
            logits=out, labels=lbl_batch))

    weights = [var for l in layers for var in l.trainable_variables]
    grads = tape.gradient(xent, weights)
    for grad, var in zip(grads, weights):
        var.assign_sub(lr*grad)

    if not step % 100:
        preds = tf.argmax(out, axis=1, output_type=tf.int32)
        acc = tf.reduce_mean(tf.cast(tf.equal(preds, lbl_batch), tf.float32))
        print("Loss: {} Accuracy: {}".format(xent, acc))


out = mnist.test_data
for layer in layers:
    out = layer(out)
test_preds = tf.argmax(out, axis=1, output_type=tf.int32)
acc = tf.reduce_mean(tf.cast(tf.equal(test_preds, mnist.test_labels), tf.float32))
print("Final test accuracy: {}".format(acc))


Loss: 2.3025596141815186 Accuracy: 0.125
Loss: 2.2998404502868652 Accuracy: 0.125
Loss: 2.296457290649414 Accuracy: 0.12890625
Starting new epoch...
Loss: 2.299673318862915 Accuracy: 0.12109375
Loss: 2.296933650970459 Accuracy: 0.125
Starting new epoch...
Loss: 2.3065595626831055 Accuracy: 0.07421875
Loss: 2.2516214847564697 Accuracy: 0.203125
Loss: 1.9676458835601807 Accuracy: 0.2734375
Starting new epoch...
Loss: 1.556753396987915 Accuracy: 0.46484375
Loss: 1.1873114109039307 Accuracy: 0.56640625
Starting new epoch...
Loss: 1.2688671350479126 Accuracy: 0.51171875
Loss: 1.0767525434494019 Accuracy: 0.609375
Starting new epoch...
Loss: 0.8752169609069824 Accuracy: 0.71875
Loss: 0.9499244689941406 Accuracy: 0.65234375
Loss: 0.8107883930206299 Accuracy: 0.75
Starting new epoch...
Loss: 0.8284597396850586 Accuracy: 0.6875
Loss: 0.9085403084754944 Accuracy: 0.66796875
Starting new epoch...
Loss: 0.8686662912368774 Accuracy: 0.7109375
Loss: 0.6725299954414368 Accuracy: 0.76171875
Starting n

In [0]:
# then load/run tensorboard

%load_ext tensorboard
%tensorboard --logdir logs

# FAIL_04

#### Observations
* during learning, due to the noise added images are little distorted

#### Posssible_fixes
* changing standard deviation(stddev) to 1 or 0.1 results in test acc of 91.4% and 93.4% each








### task_05

In [0]:
import tensorflow as tf
import datetime
from datasets import MNISTDataset
from time import time


# get the data
(train_imgs, train_lbls), (test_imgs, test_lbls) = tf.keras.datasets.mnist.load_data()
mnist = MNISTDataset(train_imgs.reshape((-1, 784)), train_lbls,
                     test_imgs.reshape((-1, 784)), test_lbls,
                     batch_size=256, seed=int(time()))


# define the model first, from input to output

# let's use fewer layers...
n_units = 100
n_layers = 2

# just set up a "chain" of hidden layers
layers = []
for layer in range(n_layers):
    layers.append(tf.keras.layers.Dense(
        n_units, activation=tf.nn.relu,
        kernel_initializer=tf.initializers.RandomUniform(minval=-0.01,
                                                         maxval=0.01),
        bias_initializer=tf.initializers.constant(0.001)))

# finally add the softmax output layer :))
layers.append(tf.keras.layers.Dense(
    10, 
    kernel_initializer=tf.initializers.RandomUniform(minval=-0.01,
                                                     maxval=0.01)))

lr = 0.1
for step in range(2000):
    img_batch, lbl_batch = mnist.next_batch()
    with tf.GradientTape() as tape:
        # here we just run all the layers in sequence via a for-loop
        out = img_batch
        for layer in layers:
            out = layer(out)
        xent = tf.reduce_mean(tf.nn.sparse_softmax_cross_entropy_with_logits(
            logits=out, labels=lbl_batch))

    weights = [var for l in layers for var in l.trainable_variables]
    grads = tape.gradient(xent, weights)
    for grad, var in zip(grads, weights):
        var.assign_sub(lr*grad)

    if not step % 100:
        preds = tf.argmax(out, axis=1, output_type=tf.int32)
        acc = tf.reduce_mean(tf.cast(tf.equal(preds, lbl_batch), tf.float32))
        print("Loss: {} Accuracy: {}".format(xent, acc))


out = mnist.test_data
for layer in layers:
    out = layer(out)
test_preds = tf.argmax(out, axis=1, output_type=tf.int32)
acc = tf.reduce_mean(tf.cast(tf.equal(test_preds, mnist.test_labels), tf.float32))
print("Final test accuracy: {}".format(acc))


Loss: 2.3026039600372314 Accuracy: 0.04296875
Loss: 2.299373149871826 Accuracy: 0.125
Loss: 2.299558162689209 Accuracy: 0.10546875
Starting new epoch...
Loss: 2.3000168800354004 Accuracy: 0.09765625
Loss: 2.2941689491271973 Accuracy: 0.1015625
Starting new epoch...
Loss: 2.0937845706939697 Accuracy: 0.19140625
Loss: 1.5501751899719238 Accuracy: 0.47265625
Loss: 0.9017661809921265 Accuracy: 0.6640625
Starting new epoch...
Loss: 0.6427043080329895 Accuracy: 0.796875
Loss: 0.6296447515487671 Accuracy: 0.796875
Starting new epoch...
Loss: 0.5869194865226746 Accuracy: 0.82421875
Loss: 0.5490639209747314 Accuracy: 0.84765625
Starting new epoch...
Loss: 0.5242233276367188 Accuracy: 0.8515625
Loss: 0.4126501679420471 Accuracy: 0.8671875
Loss: 0.3858996033668518 Accuracy: 0.8984375
Starting new epoch...
Loss: 0.3020119071006775 Accuracy: 0.921875
Loss: 0.2693336606025696 Accuracy: 0.93359375
Starting new epoch...
Loss: 0.33156073093414307 Accuracy: 0.890625
Loss: 0.35210201144218445 Accuracy: 0

In [0]:
# then load/run tensorboard

%load_ext tensorboard
%tensorboard --logdir logs

The tensorboard extension is already loaded. To reload it, use:
  %reload_ext tensorboard


Reusing TensorBoard on port 6006 (pid 32648), started 14:50:00 ago. (Use '!kill 32648' to kill it.)

# FAIL_05


#### Observations
* The gradient becomes very less as the softmax activation is applied twice on the logits. This diminishes the benefit of -ve log likelihood loss and saturation occurs in the output.



layers.append(tf.keras.layers.Dense(
    10, activation=tf.nn.softmax,
    kernel_initializer=tf.initializers.RandomUniform(minval=-0.01,
                                                     maxval=0.01)))
* due to this test acc is just above 11%

#### Possible_changes
* Remove the softmax activation at output layer, this willl lead to a test acc of 92.9%





# An MLP training script using tf.data



In [0]:
import tensorflow as tf

import numpy as np
import matplotlib.pyplot as plt

#from datasets import MNISTDataset

In [0]:
mnist = tf.keras.datasets.mnist
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()


train_data = tf.data.Dataset.from_tensor_slices((train_images, train_labels))
test_data = tf.data.Dataset.from_tensor_slices((test_images, test_labels))



In [0]:
# things we really gotta do:
# - normalize the images to [0, 1] (first convert to float)
# - reshape images from (28, 28) to (784,) (although we could do this later!)
# - convert labels to int32 (otherwise tensorflow is gonna be sad :( )

train_images = (train_images.astype(np.float32) / 255.).reshape((-1, 784))
test_images = (test_images.astype(np.float32) / 255.).reshape((-1, 784))

train_labels = train_labels.astype(np.int32)
test_labels = test_labels.astype(np.int32)

train_data = tf.data.Dataset.from_tensor_slices((train_images, train_labels))
test_data = tf.data.Dataset.from_tensor_slices((test_images, test_labels))
#train_data = train_data.batch(128)
train_data = train_data.repeat(10)
train_data = train_data.batch(128)

train_data = train_data.shuffle(buffer_size=500)



### tf.data API eliminates the use of training steps being explicitly called, by providing .repeat() function

In [0]:
learning_rate = 0.1
W= tf.Variable(tf.random.uniform(shape = [784,50], minval=-0.1,maxval=0.1, dtype=tf.float32, seed=42))
b = tf.Variable( tf.random.uniform(shape = [50], minval=-0.1,maxval=0.1, dtype=tf.float32, seed=42))

W1= tf.Variable(tf.random.uniform(shape = [50,10], minval=-0.1,maxval=0.1, dtype=tf.float32, seed=42))
b1 =tf.Variable( tf.random.uniform(shape = [10], minval=-0.1,maxval=0.1, dtype=tf.float32, seed=42))


#### **Dataset.shuffle** doesn't signal the end of an epoch until the shuffle buffer is empty. So a shuffle placed before a repeat will show every element of one epoch before moving to the next:

In [0]:
#for step in range(2):
i =0
for img_batch, lbl_batch in train_data:
    i=i+1
    with tf.GradientTape() as tape:
        logits = tf.matmul(tf.nn.relu(tf.matmul(img_batch, W) + b),W1) +b1 

        xent = tf.reduce_mean(tf.nn.sparse_softmax_cross_entropy_with_logits(
            logits=logits, labels=lbl_batch))
        
        
    dW, db, dW1, db1 = tape.gradient(xent, [W, b, W1, b1])    

    W.assign_sub(learning_rate * dW)
    b.assign_sub(learning_rate * db)
    W1.assign_sub(learning_rate * dW1)
    b1.assign_sub(learning_rate * db1)
    
        
    if not i % 500:
      preds = tf.argmax(logits, axis=1, output_type=tf.int32)
      acc = tf.reduce_mean(tf.cast(tf.equal(preds, lbl_batch),
                             tf.float32))
      print("Loss: {} Accuracy: {}".format(xent, acc))

Loss: 0.26856154203414917 Accuracy: 0.9140625
Loss: 0.1293889582157135 Accuracy: 0.9609375
Loss: 0.35489124059677124 Accuracy: 0.9140625
Loss: 0.14832685887813568 Accuracy: 0.9609375
Loss: 0.20908045768737793 Accuracy: 0.9375
Loss: 0.2550239562988281 Accuracy: 0.90625
Loss: 0.1782987117767334 Accuracy: 0.921875
Loss: 0.0496942438185215 Accuracy: 0.9765625
Loss: 0.06334952265024185 Accuracy: 0.96875


In [0]:
test_preds = tf.argmax( tf.matmul(tf.nn.relu(tf.matmul(test_images, W) + b),W1) +b1                
                       , axis=1, output_type=tf.int32)
acc1 = tf.reduce_mean(tf.cast(tf.equal(test_preds, test_labels),
                             tf.float32))
print(acc1)

tf.Tensor(0.9622, shape=(), dtype=float32)


### Result of 6 type of Ordering
| **Order**             |**Test Accuracy**|
|-----------------------|---------------|
| Shuffle,Batch,Repeat |       95.7%    |
| Shuffle,Repeat,Batch |       96.1%   |
| Batch,Repeat,Shuffle |       96.0%   |
| Batch,Shuffle,Repeat |       95.8%   |
| Repeat,Shuffle,Batch |       96.1%   |
| Repeat,Batch,Shuffle |       96.2%   |


The optimum order of using these functions should consider these points:
- Batch of data to be trained should be picked after shuffling to aviod bias issues
- Repeat should be applied before shuffing because
 
 *.shuffle doesn't signal the end of an epoch until the shuffle buffer is empty. So a shuffle placed before a repeat will show every element of one epoch before moving to the next*

---

A more descriptive information is given here: [Repeat, Shuffle & Batch](https://stackoverflow.com/questions/53514495/what-does-batch-repeat-and-shuffle-do-with-tensorflow-dataset)

#### The best order would be 
(1) Repeat 
(2) Shuffle
(3) Batch 